federation/pkg/federation-controller/namespace/namespace_controller.go

```diff
@@ -329,73 +442,79 @@ func (nc *NamespaceController) delete(namespace *api_v1.Namespace) error {
 	if err != nil {
-		return fmt.Errorf("failed to delete service list: %v", err)
+		return nil, fmt.Errorf("failed to delete service list: %v", err)
 	}
 	err = nc.federatedApiClient.Extensions().ReplicaSets(namespace.Name).DeleteCollection(&api_v1.DeleteOptions{}, api_v1.ListOptions{})
 	if err != nil {
-		return fmt.Errorf("failed to delete replicaset list from namespace: %v", err)
+		return nil, fmt.Errorf("failed to delete replicaset list from namespace: %v", err)
 	}
 	err = nc.federatedApiClient.Core().Secrets(namespace.Name).DeleteCollection(&api_v1.DeleteOptions{}, api_v1.ListOptions{})
 	if err != nil {
-		return fmt.Errorf("failed to delete secret list from namespace: %v", err)
+		return nil, fmt.Errorf("failed to delete secret list from namespace: %v", err)
 	}
 	err = nc.federatedApiClient.Extensions().Ingresses(namespace.Name).DeleteCollection(&api_v1.DeleteOptions{}, api_v1.ListOptions{})
 	if err != nil {
-		return fmt.Errorf("failed to delete ingresses list from namespace: %v", err)
+		return nil, fmt.Errorf("failed to delete ingresses list from namespace: %v", err)
 	}
 	err = nc.federatedApiClient.Extensions().DaemonSets(namespace.Name).DeleteCollection(&api_v1.DeleteOptions{}, api_v1.ListOptions{})
 	if err != nil {
-		return fmt.Errorf("failed to delete daemonsets list from namespace: %v", err)
+		return nil, fmt.Errorf("failed to delete daemonsets list from namespace: %v", err)
 	}
 	err = nc.federatedApiClient.Extensions().Deployments(namespace.Name).DeleteCollection(&api_v1.DeleteOptions{}, api_v1.ListOptions{})
 	if err != nil {
-		return fmt.Errorf("failed to delete deployments list from namespace: %v", err)
+		return nil, fmt.Errorf("failed to delete deployments list from namespace: %v", err)
 	}
 	err = nc.federatedApiClient.Core().Events(namespace.Name).DeleteCollection(&api_v1.DeleteOptions{}, api_v1.ListOptions{})
 	if err != nil {
-		return fmt.Errorf("failed to delete events list from namespace: %v", err)
+		return nil, fmt.Errorf("failed to delete events list from namespace: %v", err)
 	}
 
-	// Remove kube_api.FinalzerKubernetes
-	if len(updatedNamespace.Spec.Finalizers) != 0 {
-		finalizerSet := sets.NewString()
-		for i := range namespace.Spec.Finalizers {
-			if namespace.Spec.Finalizers[i] != api_v1.FinalizerKubernetes {
-				finalizerSet.Insert(string(namespace.Spec.Finalizers[i]))
-			}
-		}
-		updatedNamespace.Spec.Finalizers = make([]api_v1.FinalizerName, 0, len(finalizerSet))
-		for _, value := range finalizerSet.List() {
-			updatedNamespace.Spec.Finalizers = append(updatedNamespace.Spec.Finalizers, api_v1.FinalizerName(value))
-		}
-		_, err := nc.federatedApiClient.Core().Namespaces().Finalize(updatedNamespace)
-		if err != nil {
-			return fmt.Errorf("failed to finalize namespace: %v", err)
-		}
+	// Remove kube_api.FinalizerKubernetes
+	if len(namespace.Spec.Finalizers) != 0 {
+		return nc.removeFinalizerFromSpec(namespace, api_v1.FinalizerKubernetes)
```

Comment:
> I think you should check if all resources are actually deleted to make the deletion synchronous. 

Yes this is not required right now since all deletions are immediate (there is no cascading deletion). Will need to update this before enabling cascading deletion for any of these resources.
The plan is to refactor kubernetes namespace controller code to make the namespace deletion code reusable and then use it here.
Leaving as is in this PR.